shared/contracts/BaseTemplate.sol

```diff
@@ -56,155 +62,178 @@ contract BaseTemplate is APMNamehash, IsContract {
+    function _removePermissionFromTemplate(ACL _acl, address _app, bytes32 _permission) internal {
+        _acl.revokePermission(address(this), _app, _permission);
+        _acl.removePermissionManager(_app, _permission);
     }
 
-    function _removePermission(ACL acl, address app, bytes32 permission) internal {
-        acl.revokePermission(address(this), app, permission);
-        acl.removePermissionManager(app, permission);
+    function _transferPermissionFromTemplate(ACL _acl, address _to, address _app, bytes32 _permission) internal {
+        _transferPermissionFromTemplate(_acl, _to, _to, _app, _permission);
     }
 
-    function _transferPermission(ACL acl, address to, address app, bytes32 permission) internal {
-        _transferPermission(acl, to, to, app, permission);
+    function _transferPermissionFromTemplate(ACL _acl, address _to, address _manager, address _app, bytes32 _permission) internal {
+        _acl.grantPermission(_to, _app, _permission);
+        _acl.revokePermission(address(this), _app, _permission);
+        _acl.setPermissionManager(_manager, _app, _permission);
     }
 
-    function _transferPermission(ACL acl, address to, address manager, address app, bytes32 permission) internal {
-        acl.grantPermission(to, app, permission);
-        acl.revokePermission(address(this), app, permission);
-        acl.setPermissionManager(manager, app, permission);
+    function _transferRootPermissionsFromTemplate(Kernel _dao, address _to) internal {
+        ACL _acl = ACL(_dao.acl());
+        _transferPermissionFromTemplate(_acl, _to, _dao, _dao.APP_MANAGER_ROLE());
+        _transferPermissionFromTemplate(_acl, _to, _acl, _acl.CREATE_PERMISSIONS_ROLE());
+    }
+
+    function _transferRootPermissionsFromTemplate(Kernel _dao, address _to, address _manager) internal {
+        ACL _acl = ACL(_dao.acl());
+        _transferPermissionFromTemplate(_acl, _to, _manager, _dao, _dao.APP_MANAGER_ROLE());
+        _transferPermissionFromTemplate(_acl, _to, _manager, _acl, _acl.CREATE_PERMISSIONS_ROLE());
     }
 
     /* AGENT */
 
-    function _installDefaultAgentApp(Kernel dao) internal returns (Agent) {
-        Agent agent = Agent(_installDefaultApp(dao, AGENT_APP_ID));
+    function _installDefaultAgentApp(Kernel _dao) internal returns (Agent) {
+        Agent agent = Agent(_installDefaultApp(_dao, AGENT_APP_ID));
         agent.initialize();
+        // We assume that installing the Agent app as default is in order to replace the Vault app which is
+        // normally installed as default. Thus, we are setting its ID as the Vault id that the Kernel will use.
+        _dao.setRecoveryVaultAppId(AGENT_APP_ID);
         return agent;
     }
 
-    function _installNonDefaultAgentApp(Kernel dao) internal returns (Agent) {
-        Agent agent = Agent(_installNonDefaultApp(dao, AGENT_APP_ID));
+    function _installNonDefaultAgentApp(Kernel _dao) internal returns (Agent) {
```

Comment:
Rather than two functions, does it make sense just to add the "default"-ness as an argument flag instead?

Reply:
I gave it a try and it increased the bytecode length. The `TrustTemplate` is hitting the limit (see [here](https://travis-ci.org/aragon/dao-templates/jobs/563044874)), honestly, I prefer the syntax-sugar rather than dealing with that now 😅

Reply:
Actually, the previous commit is already exceeding the limit, don't get why it passed on the CI 🤷‍♀

Reply:
Hm, interesting. I did a quick check and I see big differences:

| Contract | Size |
|-------------|-------|
| Trust | 24625 |
| Democracy | 11835 |
| Multisig | 12019 |
| Company | 12188 |
| C. Board | 15350 |
| Reputation | 12045 |
| Membership | 11953 |

Any idea why? what does this Trust one have that makes it twice as big as most of the other ones?

Reply:
I think once we optimize apps installation with a single call instead of two we will have more space for this change

Reply:
@bingen yea, probably because of the Gnosis multisig